### ISSUE TITLE
GAP limitation in ACK Frame

### ISSUE THREAD

**Issue by gloinul** [quicwg/base-drafts#613]
In draft-ietf-quic-transport-03, Section 8.2.1:

```
    0                   1                   2                   3
    0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
   |              First ACK Block Length (8/16/32/48)            ...
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
   |  [Gap 1 (8)]  |       [ACK Block 1 Length (8/16/32/48)]     ...
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
   |  [Gap 2 (8)]  |       [ACK Block 2 Length (8/16/32/48)]     ...
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
                                ...
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
   |  [Gap N (8)]  |       [ACK Block N Length (8/16/32/48)]     ...
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
```
>   Gap To Next Block (opt, repeated):  An unsigned number specifying the
      number of contiguous missing packets from the end of the previous
      ACK block to the start of the next.  Repeated "Num Blocks" times.

So, to my understanding the gap counter is fixed to 8 bits, and thus have a maximum GAP distance of 255. Considering that packet numbers are not required to be a continuous number, nor have a limit that the largest gap is less than 255, I wonder how one handles situations where the GAP between received packet numbers are larger than 255? Do one use multiple ACK frames, where each acking the sub-set of packet numbers that are separeted by GAPs larger than 255?

**Comment by mcmanus** [comment]
yes, new ack frame.

**Comment by marten-seemann** [comment]
You don't need a new frame. You can just set the length to 0 if you need to encode a gap larger than 255.

**Comment by ianswett** [comment]
Marten is correct, and at some point, as Patrick said, a new ACK frame is necessary.

I could add text to clarify this?

**Comment by gloinul** [comment]
So this needs clarification in the text on how you handle the case I brought up. I also see a relation to this issue  #423. So please add text on this.

**Comment by ianswett** [comment]
I sent out a PR for clarifying the use of 0 length ack ranges, but I think it's best to resolve the multiple ack frames issue on #423

**Comment by phluid61** [comment]
Is this the same (or worse) for timestamps in the ACK frame?

The current text is pretty vague, but seems to say that the _Delta Largest Acked 1..N_ value is always at most 255 from the _Largest Acknowledged_ (not the previous delta).

So... timestamps can go backwards (or rather, packet numbers can go backwards inside the timestamp section), and you can't timestamp packets with much larger numbers.

Even if the _Delta N_ was changed to be relative to _Delta N-1_, you still couldn't set the _Time Since Previous_ to 0 to skip over gaps>255, since the text says that the delta is used to calculate the number of the actual packet being timestamped.

**Comment by ianswett** [comment]
Timestamps are only sent once, so assuming there isn't a large amount of reordering or loss, up the the 255th most recent is more than sufficient.

But yes, if you're commonly experiencing loss and reordering >255 packets, you won't get timestamps.  But that is extraordinarily rare in my experience, and since timestamps are best effort, one should use them as an enhancement to congestion control, not the sole input.

**Comment by ianswett** [comment]
Closed by #595 and #841

**Comment by ianswett** [comment]
The issue was #595, but the PR was #877
